code/game/machinery/computer/launchpad_control.dm

```diff
@@ -12,6 +12,102 @@
 	. = ..()
+	AddComponent(/datum/component/usb_port, list(
+		/obj/item/circuit_component/bluespace_launchpad,
+	))
+
+/obj/item/circuit_component/bluespace_launchpad
+	display_name = "Bluespace Launchpad Console"
+	display_desc = "Teleports anything to and from any location on the station. Doesn't use actual GPS coordinates, but rather offsets from the launchpad itself. And it can only go as far as the launchpad can go, which depends on its components."
+
+	var/datum/port/input/launchpad_id
+	var/datum/port/input/x_pos
+	var/datum/port/input/y_pos
+	var/datum/port/input/send_trigger
+	var/datum/port/input/retrieve_trigger
+
+	var/datum/port/output/sent
+	var/datum/port/output/retrieved
+	var/datum/port/output/on_fail
+	var/datum/port/output/why_fail
+
+	var/obj/machinery/computer/launchpad/attached_console
+
+/obj/item/circuit_component/bluespace_launchpad/Initialize()
+	. = ..()
+	launchpad_id = add_input_port("Launchpad ID", PORT_TYPE_NUMBER, FALSE)
+	x_pos = add_input_port("X offset", PORT_TYPE_NUMBER, FALSE)
+	y_pos = add_input_port("Y offset", PORT_TYPE_NUMBER, FALSE)
+	send_trigger = add_input_port("Send", PORT_TYPE_SIGNAL)
+	retrieve_trigger = add_input_port("Retrieve", PORT_TYPE_SIGNAL)
+
+	sent = add_output_port("Sent", PORT_TYPE_SIGNAL)
+	retrieved = add_output_port("Retrieved", PORT_TYPE_SIGNAL)
+	why_fail = add_output_port("Fail reason", PORT_TYPE_STRING)
+	on_fail = add_output_port("Failed", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/bluespace_launchpad/Destroy()
+	launchpad_id = null
+	x_pos = null
+	y_pos = null
+	send_trigger = null
+	retrieve_trigger = null
+	sent = null
+	retrieved = null
+	on_fail = null
+	why_fail = null
+	return ..()
+
+/obj/item/circuit_component/bluespace_launchpad/register_usb_parent(atom/movable/parent)
+	. = ..()
+	if(istype(parent, /obj/machinery/computer/launchpad))
+		attached_console = parent
+
+/obj/item/circuit_component/bluespace_launchpad/unregister_usb_parent(atom/movable/parent)
+	attached_console = null
+	return ..()
+
+/obj/item/circuit_component/bluespace_launchpad/input_received(datum/port/input/port)
+	. = ..()
+
+	if(.)
+		return
+
+	if(!attached_console || length(attached_console.launchpads) == 0)
+		on_fail.set_output(COMPONENT_SIGNAL)
+		why_fail.set_output("No launchpads connected!")
+		return
+
+	var/current_launchpad = launchpad_id.input_value
+	if(isnull(current_launchpad))
+		return
+
+	var/obj/machinery/launchpad/the_pad = attached_console.launchpads[current_launchpad]
+	if(isnull(the_pad))
+		return
+
+	var/x_dest = x_pos.input_value
+	if(isnull(x_dest))
+		return
+
+	var/y_dest = y_pos.input_value
+	if(isnull(y_dest))
+		return
+
+	if(x_dest > the_pad.range || y_dest > the_pad.range)
+		on_fail.set_output(COMPONENT_SIGNAL)
+		why_fail.set_output("Cannot go that far! Current maximum reach: [the_pad.range]")
+		return
+
+	the_pad.set_offset(x_dest, y_dest)
+
+	if(COMPONENT_TRIGGERED_BY(send_trigger, port))
+		attached_console.teleport(usr, the_pad, TRUE)
```

Comment:
`usr` has a few problems here.

1. `usr` is hard to predict from reading the code of who it is actually refering to.
2. `launchpad.teleport` sends `to_chat`s to the user it is given. This means the user could be very far away yet still get to_chat's. Ideally they would not get any.

Here's what you need to do

1. Turn the checks in `teleport` (and the proc it calls) into its own thing. That means instead of printing `"ERROR: Launchpad busy."` to the user's chat, it's instead a return value. The current code can just `to_chat` this like it does now, but circuit code would ideally pass it through the `why_fail` node you have now.
2. `user` is also used for logging who teleported what. You would still want this here (a big problem with old circuits was lack of logging), except you'd want to be like arrest consoles where they get it based on whoever screwed it in, etc.

Reply:
usr for circuits is also being disabled.
They will soon be unusable in input_received() procs, so do be aware of that.